Customizing typing indicator (#2912)
microsoft/BotFramework-WebChat: docs/HOOKS.md

-  [`useVoiceSelector`](#usevoiceselector)
-  [`useWebSpeechPonyfill`](#usewebspeechponyfill)

## `useActiveTyping`

```js
interface Typing {
  at: number;
  expireAt: number;
  name: string;
  role: 'bot' | 'user';
}

useActiveTyping(expireAfter?: number): [{ [id: string]: Typing }]
```

This function will return a list of participants who are actively typing, including the start typing time (`at`) and expiration time (`expireAt`), the name and the role of the participant.

> If the participant send a message after the typing activity, the participant will be explicitly removed from the list. If no message or typing activity is received, the participant is considered inactive and not listed in the result. To keep typing indicator active, participants should send typing activity continuously.

Reviewer: ```suggestion
If the participant sends a message after the typing activity, the participant will be explicitly removed from the list. If no messages or typing activities are received, the participant is considered inactive and not listed in the result. To keep the typing indicator active, participants should continuously send the typing activity.
```

> The inactivity timer can be overridden by passing the `expireAfter` argument. If `Infinity` is passed, it will return all participants who did not explicitly remove from the list. In other words, it will return participants who sent typing activity, but did not send a message activity afterward.

Reviewer: ```suggestion
The `expireAfter` argument can override the inactivity timer. If `expireAfter` is `Infinity`, it will return all participants who did not explicitly remove from the list. In other words, it will return participants who sent a typing activity, but did not send a message activity afterward.
```

Author: Love the reversion of words makes the sentence much cleaner.

To modify this value, change the value in the style options prop passed to Web Chat.

## `useRenderToast`

```js
interface Notification {
  alt?: string;
  id: string;
  level: 'error' | 'info' | 'success' | 'warn' | string;
  message: string;
}

useRenderToast(): ({ notification: Notification }) => React.Element
```

> This function is for rendering a toast for the notification toaster. The caller will need to pass `notification` as parameter. This function is a composition of `toastMiddleware`, which is passed as a prop.

Reviewer: ```suggestion
This function is for rendering a toast for the notification toaster. The caller will need to pass `notification` as parameter to the function. This function is a composition of `toastMiddleware`, which is passed as a prop to Web Chat.
```

## `useRenderTypingIndicator`

```js
interface Typing {
  at: number;
  expireAt: number;
  name: string;
  role: 'bot' | 'user';
}

useRenderTypingIndicator():
  ({
    activeTyping: { [id: string]: Typing },
    typing: { [id: string]: Typing },
    visible: boolean
  }) => React.Element
```

> This function is for rendering typing indicator for all participants. This function is a composition of `typingIndicatorMiddleware`, which is passed as a prop. The caller will pass the following arguments:

Reviewer: ```suggestion
This function is for rendering typing indicator for all participants of the conversation. This function is a composition of `typingIndicatorMiddleware`, which is passed as a prop to Web Chat. The caller will pass the following arguments:
```

> -  `activeTyping` lists participants who are actively typing.

Reviewer: ```suggestion
-  `activeTyping` lists of participants who are actively typing.
```
